Check feature flag on backend /graph PUT API for block permission

jeff303 opened the pull request:
Check for the `:advanced-permissions` flag in API handler, when a `:schemas` `:block` entry is present in the graph

Set feature flag in tests when using API

Add new test to ensure failure in the case that feature flag is missing

jeff303 committed:
Check feature flag on backend /graph PUT API for block permission

Check for the `:advanced-permissions` flag in API handler, when a `:schemas` `:block` entry is present in the graph

Set feature flag in tests when using API

Add new test to ensure failure in the case that feature flag is missing

camsaul commented on enterprise/backend/test/metabase_enterprise/enhancements/models/permissions/block_permissions_test.clj:
```diff
@@ -139,7 +152,8 @@
                         s/Keyword s/Any}
-                       (mt/user-http-request :crowberto :put 500 "permissions/graph" new-graph))))))))
+                       (premium-features-test/with-premium-features #{:advanced-permissions}
+                         (mt/user-http-request :crowberto :put 500 "permissions/graph" new-graph)))))))))
```
This is not part of this PR but I think we probably should be returning a 400 instead of a 500 here

jeff303 replied:
Yeah, let's tackle that in a separate PR.

camsaul commented on src/metabase/api/permissions.clj:
```diff
@@ -46,6 +48,16 @@
+                         x)) graph)
+      (when (and @has-block-perms? (not (premium-features/has-feature? :advanced-permissions)))
+        (throw (ex-info (tru "Can''t use block permissions without having the advanced-permissions premium feature")
+                        {:status-code 403}))))
```
I've been using [`402 Payment Required`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/402) for the other feature-flagged responses. Not 100% sure it makes sense but it does at least make the tests easier to write because we can confirm something is failing because of token feature flags rather than perms

camsaul replied:
In my impl for the fix I just added this check to `update-db-permissions!` in the `:block` code... that way you don't have to do any walking

jeff303 replied:
Fixed both

camsaul reviewed (approved):
This fix works for me. I think you could make it a little simpler by doing the check inside `update-db-permissions!` when you see `:block` so you don't have to walk the whole thing.

It would also be nice to check for `:segmented` permissions and the `:sandboxes` feature -- but that's always worked that way so fixing it before the new release isn't as much of a priority

jeff303 committed:
PR feedback

Put check into `update-db-permissions!` and remove the walk

Change status code to 402

jeff303 committed:
Fix test failures, hopefully
